simulator/infra/memory/t/unit_test.cpp

```diff
@@ -57,6 +61,58 @@ TEST_CASE( "Func_memory: StartPC_Method_Test")
     CHECK( func_mem.startPC() == 0x4000b0u /*address of the ".text" section*/);
 }
 
+TEST_CASE( "Func_memory: Host_Guest_Memcpy_1b")
+{
+    // Single byte
+    TestFuncMemory func_mem;
+    const Byte write_data = Byte( 0xA5);
+    Byte read_data = Byte( 0xFF);
+
+    CHECK( func_mem.memcpy_host_to_guest_noexcept( dataSectAddr, &write_data, 1) == 1);
+    CHECK( func_mem.check_and_read_byte( dataSectAddr) == write_data);
```

Comment:
I do not understand why we should test that method, as we have memcpy as an external interface.

Reply:
> check_and_read_byte() is used to check if bytes written by 

Yes, and the same is made by `memcpy_host_to_guest`.

> memcpy's are not external only, they are used by read() and write(), which are used in FuncSim memory stage.

All of these functions are external to FuncMemory class.
`check_and_read_byte` is not external.

Reply:
I don't understand the question. We don't need to test memcpy, right?

Reply:
> I don't understand the question.

I did not ask any question :-)

> We don't need to test memcpy, right?

We need and we do now.

```c++
TEST_CASE( "Func_memory: Host_Guest_Memcpy_1b")	
{	
    // Arranging: memory, byte to write, byte to save
    FuncMemory func_mem;	
    const Byte write_data_1 = Byte( 0xA5);	
    Byte read_data_1 = Byte( 0xFF);	

    // Act: write a byte into memory
    CHECK( func_mem.memcpy_host_to_guest_noexcept( dataSectAddr, &write_data_1, 1) == 1);	

    // Assert: the same byte is obtained from the memory.
    CHECK( func_mem.read<uint8>( dataSectAddr) == static_cast<uint8>( write_data_1));	
    CHECK( func_mem.memcpy_guest_to_host_noexcept( &read_data_1, dataSectAddr, 1) == 1);	
    CHECK( read_data_1 == write_data_1);	
 }
```

Reply:
Now I get it. What if both `host_to_guest` and `guest_to_host` have errors (e.g, swap high and low 4 bits), and  `guest_to_host` gives correct result (0xA5), but actual byte in memory is 0x5A? So I decided to check memory directly with internal method `check_and_read_byte()`.

Reply:
> What if both host_to_guest and guest_to_host have errors (e.g, swap high and low 4 bits), and guest_to_host gives correct result (0xA5), but actual byte in memory is 0x5A? 

Is it really a problem? It works as we expect, and we do not care about the internals. Values may be transmitted to reptilians' UFO and back, I do not care.

Reply:
Moreover, that is the advantage. If someone reveals that it is faster/power-saving/cheaper to keep values with digits swapped (or, more, realistic, keep the bits inverted), we can easily change our implementation so no one would notice the change.